lib/pbench/test/unit/server/query_apis/commons.py

```diff
@@ -0,0 +1,198 @@
+            json={"user": user, "start": "2020-08", "end": "2020-10"},
+        )
+        assert response.status_code == HTTPStatus.FORBIDDEN
+
+    def test_missing_json_object(self, client, server_config, pbench_token):
+        """
+        Test behavior when no JSON payload is given
+        """
+        response = client.post(
+            f"{server_config.rest_uri}{self.endpoint}",
+            headers={"Authorization": "Bearer " + pbench_token},
+        )
+        assert response.status_code == HTTPStatus.BAD_REQUEST
+        assert response.json.get("message") == "Invalid request payload"
+
+    def test_missing_keys(self, client, server_config, keys, user_ok, pbench_token):
+        """
+        Test behavior when JSON payload does not contain all required keys.
+
+        Note that "start", and "end" are required whereas "user" is not mandatory;
+        however, Pbench will silently ignore any additional keys that are
+        specified.
+       """
+        response = client.post(
+            f"{server_config.rest_uri}{self.endpoint}",
+            headers={"Authorization": "Bearer " + pbench_token},
+            json=keys,
+        )
+        assert response.status_code == HTTPStatus.BAD_REQUEST
+        missing = [k for k in self.required_keys if k not in keys]
+        assert (
+            response.json.get("message")
+            == f"Missing required parameters: {','.join(missing)}"
+        )
+
+    def test_bad_dates(self, client, server_config, user_ok, pbench_token):
+        """
+        Test behavior when a bad date string is given
+        """
+        response = client.post(
+            f"{server_config.rest_uri}{self.endpoint}",
+            headers={"Authorization": "Bearer " + pbench_token},
+            json=self.json_bad_date,
+        )
+        assert response.status_code == HTTPStatus.BAD_REQUEST
+        assert (
+            response.json.get("message")
+            == "Value '2020-19' (str) cannot be parsed as a date/time string"
+        )
+
+    def test_empty_query(
+        self,
+        client,
+        server_config,
+        query_api,
+        user_ok,
+        find_template,
+        build_auth_header,
+    ):
+        """
+        Check proper handling of a query resulting in no Elasticsearch matches.
+        The test will run thrice with different values of the build_auth_header
+        fixture.
+        """
+        try:
+            self.empty_response_payload
+        except AttributeError:
+            pytest.skip(
+                "skipping the empty query test since the empty response payload is not set"
+            )
+        expected_status = HTTPStatus.OK
+        if build_auth_header["header_param"] != "valid":
+            expected_status = HTTPStatus.FORBIDDEN
+
+        index = self.build_index(server_config, ("2020-08", "2020-09", "2020-10"))
+        response = query_api(
+            self.endpoint,
+            "/_search?ignore_unavailable=true",
```

Comment:
This will run for each subclass test, but it only makes sense for the APIs built on top of `_search`. (E.g., not publish, or months, or index templates.) Maybe you can work out a way for the subclasses to configure the Elastic endpoint and response payload?